opentelemetry-sdk/src/opentelemetry/sdk/metrics/_internal/aggregation.py

```diff
@@ -349,6 +362,326 @@ def collect(
+        self._sum = 0
+
+        # _count is the count of all calls to aggregate.
+        self._count = 0
+
+        # _zero_count is the count of all the calls to aggregate when the value
+        # to be aggregated is exactly 0.
+        self._zero_count = 0
+
+        # _min is the smallest value aggregated by this aggregator.
+        self._min = inf
+
+        # _max is the smallest value aggregated by this aggregator.
+        self._max = -inf
+
+        # _positive holds the positive values.
+        self._positive = Buckets()
+
+        # _negative holds the negative values by their absolute value.
+        self._negative = Buckets()
+
+        # _mapping corresponds to the current scale, is shared by both the
+        # positive and negative buckets.
+        self._mapping = LogarithmMapping(LogarithmMapping._max_scale)
+
+        self._instrument_temporality = AggregationTemporality.DELTA
+        self._start_time_unix_nano = start_time_unix_nano
+
+    def aggregate(self, measurement: Measurement) -> None:
+        # pylint: disable=too-many-branches,too-many-statements
+
+        value = measurement.value
+
+        # 0. Set the following attributes:
+        # _min
+        # _max
+        # _count
+        # _zero_count
+        # _sum
+        if value < self._min:
+            self._min = value
+
+        if value > self._max:
+            self._max = value
+
+        self._count += 1
+
+        if value == 0:
+            self._zero_count += 1
+            return
+
+        self._sum += value
+
+        # 1. Use the positive buckets for positive values and the negative
+        # buckets for negative values.
+        if value > 0:
+            buckets = self._positive
+
+        else:
+            # Both exponential and logarithm mappings use only positive values
+            # so the absolute value is used here.
+            value = -value
+            buckets = self._negative
+
+        # 2. Compute the index for the value at the current scale.
+        index = self._mapping.map_to_index(value)
+
+        # 3. Determine if a change of scale is needed.
+        is_rescaling_needed = False
+
+        if len(buckets) == 0:
+            buckets._index_start = index
+            buckets._index_end = index
+            buckets._index_base = index
+
+        elif (
+            index < buckets._index_start
+            and (buckets._index_end - index) >= self._max_size
+        ):
+            is_rescaling_needed = True
+            low = index
+            high = buckets._index_end
+
+        elif (
+            index > buckets._index_end
+            and (index - buckets._index_start) >= self._max_size
+        ):
+            is_rescaling_needed = True
+            low = buckets._index_start
+            high = index
+
+        # 4. Rescale the mapping if needed.
+        if is_rescaling_needed:
+
+            change = 0
+
+            while high - low >= self._max_size:
```

Comment:
Nit: This "new scale" algorithm is dying to be called out as a helper method.  Something like:

```python
def get_scale_reduction(start, end, current):
```

defined directly on `Buckets`.

Right now, this code is harder to follow and would really benefit from some abstraction to simplify responsibility.  There's a lot of code here "reaching into" the Buckets class.

Reply:
That rescaling algorithm is used in our code only in one place, I prefer not to have "single-use functions" because that makes the code hard to read, the reader has to make a mental note of the values passed to the function, scroll up or down to find the function (or find it in another file), paste the values, read the function code and come back to the place where the function call was.

Reply:
Hm, I think I was confused by the line mentioned above, I see now how we are setting private buckets values like `_index_start` and `_index_end` in this code, it may be possible to make a new function for the `Buckets` class that does this, as you mention :+1:

Reply:
After taking another look, I have refactored this by adding properties for the buckets attributes accessed here. I prefer not to create a function for this because it is very hard to find a descriptive name for it.